[TDF] Report returns an object to inspect efficiencies of cuts programmatically

dpiparo opened the pull request.

bluehood commented on tree/treeplayer/inc/ROOT/TDFInterface.hxx:
```diff
@@ -751,8 +751,10 @@ public:
-   /// \param[in] end Final entry number (excluded) considered for this range. 0 means that the range goes until the end of the dataset.
-   /// \param[in] stride Process one entry of the [begin, end) range every `stride` entries. Must be strictly greater than 0.
+   /// \param[in] end Final entry number (excluded) considered for this range. 0 means that the range goes until the end
+   /// of the dataset.
```
don't formatting changes like this break the doxygen doc?

dpiparo replied:
arg. I thought we protected that. Thanks for spotting this.

bluehood commented on tree/treeplayer/inc/ROOT/TDFNodes.hxx:
```diff
@@ -31,6 +31,50 @@
+   float GetEff() const;
+};
+
+class TCutFlowReport {
```
Maybe `TCutFlowReport` is important enough to deserve its own file

bluehood commented on tree/treeplayer/src/TDFNodes.cxx:
```diff
@@ -29,6 +29,67 @@ using namespace ROOT::Detail::TDF;
+ULong64_t TCutInfo::GetPass() const
+{
+   return fPass;
+}
```
we usually leave these inlined in the header?

bluehood commented on tree/treeplayer/src/TDFNodes.cxx:
```diff
@@ -29,6 +29,67 @@ using namespace ROOT::Detail::TDF;
+{
+   if (cutName.empty()) {
+      throw std::runtime_error("Cannot look for an unnamed cut.");
+   }
```
this special case would be already handled by the following logic

bluehood commented on tree/treeplayer/inc/ROOT/TDFInterface.hxx:
```diff
@@ -1495,21 +1498,27 @@ public:
    /// `TDataFrame` and that node (included). Stats are printed in the same
    /// order as the named filters have been added to the graph.
-   void Report()
+   TCutFlowReport Report()
```
I think it would be nice if we took a flag that users can toggle to decide whether they want the cuts printed on screen or not.

dpiparo commented:
Thanks @bluehood for the constructive comments. I implemented them and added another feature, i.e. the possibility not to print based on a boolean flag passed to report.
I can easily implement the value printer for the TCutFlowReport once it is decided how we want to better support interactive usage.

bluehood commented on tree/treeplayer/inc/ROOT/TCutFlowReport.hxx:
```diff
@@ -0,0 +1,67 @@
+namespace Experimental {
+namespace TDF {
+
+class TCutFlowReport;
```
no need for this, friend declarations already forward-declare objects in the same namespace
